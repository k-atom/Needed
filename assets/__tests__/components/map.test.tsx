import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { mount, ReactWrapper } from 'enzyme';

import MapContainer from '../../ts/components/map';

let lat: number = 25.032697;
let lng: number = 121.563086;
let position:[number, number] = [lat, lng];
let zoom: number = 13;

const setLat = (num: number) => {
  lat = num;
  setPosition([lat, lng]);
};

const setLng = (num: number) => {
  lng = num;
  setPosition([lat, lng]);
};

const setPosition = ([lat, lng]: [number, number]) => {
  position = [lat, lng];
  previous.position = position;
};

const setZoom = (num: number) => {
  zoom = num;
};

let previous = {
  lat: lat,
  lng: lng,
  position: position,
  zoom: zoom,
  fn: {
    setLat: setLat,
    setLng: setLng,
    setPosition: setPosition,
    setZoom: setZoom
  },
  MapControlRef: {}
};

/**
 * This function resolves the 'act(...)' warnings
 * during the test.
 *
 * The warning seems to be triggered by asynchronous
 * calls to setData triggered by a useEffect function.
 * @param {ReactWrapper} wrapper A ReactWrapper to do update
 * @param {() => void)} _actions This function to do call back
 */
const actions = async (wrapper: ReactWrapper, _actions: any) => {
  await act(async () => {
    await (new Promise(resolve => setTimeout(resolve, 0)));
    _actions();
    wrapper.update();
  });
};

function SnapContainer() {
  previous.MapControlRef = React.useRef();

  return (
    <>
      <MapContainer {...previous} />
    </>
  );
}

describe('MapContainer', () => {
  const component = mount(<SnapContainer />);

  it('Check node of MapContainer', () => {
    let componentRender = component.render();

    expect(component.find('Map')).toHaveLength(1);
    expect(componentRender.find('div.leaflet-container')).toHaveLength(1);
    expect(componentRender.find('div.mapComponent')).toHaveLength(1);
    expect(componentRender.find('a.leaflet-control-zoom-in')).toHaveLength(1);
    expect(componentRender.find('a.leaflet-control-zoom-out')).toHaveLength(1);
    expect(component.html()).toMatchSnapshot();
  });
  it('Node of MapContainer changes when previous position changes', () => {
    let snapHTML: string;

    snapHTML = component.html();
    previous.fn.setLat(0);
    component.setProps(previous);
    expect(component.html() != snapHTML ).toBe(true);

    snapHTML = component.html();
    previous.fn.setLng(0);
    component.setProps(previous);
    expect(component.html() != snapHTML ).toBe(true);

    snapHTML = component.html();
    previous.fn.setPosition([1, 1]);
    component.setProps(previous);
    expect(component.html() != snapHTML ).toBe(true);
  });
  it('Should successfully call onViewportChange of Leaflet', async () => {
    await actions(component, () => {
      let props: any = component.find('Map').props();

      props.onViewportChange({zoom: 12});
    });
  });
  it('Should successfully call onCreated of EditControlFeatureGroup', async () => {
    await actions(component, () => {
      let props: any = component.find('EditControlFeatureGroup').props();
      let evt = {
        layer: {
          getLatLng: () => {
            return {
              lat: 25,
              lng: 121
            };
          }
        }
      };

      props.onCreated(evt);
    });
  });
});
