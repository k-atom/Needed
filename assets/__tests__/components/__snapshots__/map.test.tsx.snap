// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapContainer Check node of MapContainer 1`] = `"<div class=\\"mapContainer\\"><div class=\\"leaflet-container leaflet-grab leaflet-touch-drag\\" style=\\"position: relative;\\" tabindex=\\"0\\"><div class=\\"leaflet-pane leaflet-map-pane\\" style=\\"left: 0px; top: 0px;\\"><div class=\\"leaflet-pane leaflet-tile-pane\\"><div class=\\"leaflet-layer \\" style=\\"z-index: 1;\\"><div class=\\"leaflet-tile-container leaflet-zoom-animated\\" style=\\"z-index: 19; left: 0px; top: 0px;\\"><img alt=\\"\\" role=\\"presentation\\" src=\\"https://b.tile.openstreetmap.org/13/6862/3507.png\\" class=\\"leaflet-tile\\" style=\\"width: 256px; height: 256px; left: -60px; top: -84px;\\"></div></div></div><div class=\\"leaflet-pane leaflet-shadow-pane\\"><img src=\\"marker-shadow.png\\" class=\\"leaflet-marker-shadow leaflet-zoom-hide\\" style=\\"margin-left: -12px; margin-top: -41px; width: 41px; height: 41px; left: 0px; top: 0px;\\" alt=\\"\\"></div><div class=\\"leaflet-pane leaflet-overlay-pane\\"></div><div class=\\"leaflet-pane leaflet-marker-pane\\"><img src=\\"marker-icon.png\\" class=\\"leaflet-marker-icon leaflet-zoom-hide leaflet-interactive\\" style=\\"margin-left: -12px; margin-top: -41px; width: 25px; height: 41px; left: 0px; top: 0px; z-index: 0;\\" alt=\\"\\" tabindex=\\"0\\"></div><div class=\\"leaflet-pane leaflet-tooltip-pane\\"></div><div class=\\"leaflet-pane leaflet-popup-pane\\"></div></div><div class=\\"leaflet-control-container\\"><div class=\\"leaflet-top leaflet-left\\"><div class=\\"leaflet-control-zoom leaflet-bar leaflet-control\\"><a class=\\"leaflet-control-zoom-in\\" href=\\"#\\" title=\\"Zoom in\\" role=\\"button\\" aria-label=\\"Zoom in\\">+</a><a class=\\"leaflet-control-zoom-out\\" href=\\"#\\" title=\\"Zoom out\\" role=\\"button\\" aria-label=\\"Zoom out\\">−</a></div><div class=\\"leaflet-draw leaflet-control\\"></div></div><div class=\\"leaflet-top leaflet-right\\"></div><div class=\\"leaflet-bottom leaflet-left\\"></div><div class=\\"leaflet-bottom leaflet-right\\"><div class=\\"leaflet-control-attribution leaflet-control\\"><a href=\\"https://leafletjs.com\\" title=\\"A JS library for interactive maps\\">Leaflet</a> | © <a href=\\"http://osm.org/copyright\\">OpenStreetMap</a> contributors</div></div></div><div class=\\"mapComponent bottom right\\"><button title=\\"定位當前位置\\" type=\\"button\\" class=\\"ant-btn btn\\"><i class=\\"material-icons\\">my_location</i></button></div></div></div>"`;
